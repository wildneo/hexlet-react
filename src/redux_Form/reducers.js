import { combineReducers } from 'redux';

const text = (state = '', action) => {
  switch (action.type) {
    case 'TEXT_CHANGE':
      return action.payload.value;

    case 'TEXT_RESET': {
      return '';
    }

    default:
      return state;
  }
};

export default combineReducers({
  text,
});
