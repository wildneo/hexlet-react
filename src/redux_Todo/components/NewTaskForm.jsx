import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import * as actions from '../actions';

const mapStateToProps = ({ text }) => ({ text });
const actionsList = {
  addTask: actions.addTask,
  updateNewTaskText: actions.updateNewTaskText,
};

class NewTaskForm extends React.Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleUpdate = this.handleUpdate.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    const { addTask, text } = this.props;
    const task = { id: _.uniqueId(), text, state: 'active' };
    addTask({ task });
  }

  handleUpdate(event) {
    event.preventDefault();
    const { updateNewTaskText } = this.props;
    const { target: { value } } = event;
    updateNewTaskText({ value });
  }

  render() {
    const { text } = this.props;
    return (
      <form action="" className="form-inline" onSubmit={this.handleSubmit}>
        <div className="form-group mx-sm-3">
          <input type="text" required value={text} onChange={this.handleUpdate} />
        </div>
        <input type="submit" className="btn btn-primary btn-sm" value="Add" />
      </form>
    );
  }
}

NewTaskForm.defaultProps = {
  text: '',
};

export default connect(mapStateToProps, actionsList)(NewTaskForm);
