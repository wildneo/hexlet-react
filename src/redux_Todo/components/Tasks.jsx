/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { connect } from 'react-redux';
import cn from 'classnames';
import * as actions from '../actions';

const mapStateToProps = (state) => {
  const { tasks: { byId, allIds }, tasksUIState } = state;
  const tasks = allIds.map((id) => byId[id]);
  return { tasks, tasksUIState };
};
const actionsList = {
  inverseTaskTheme: actions.inverseTaskTheme,
  toggleTaskState: actions.toggleTaskState,
  removeTask: actions.removeTask,
};

class Tasks extends React.Component {
  constructor(props) {
    super(props);

    this.handleToggleTaskState = (id) => (event) => {
      event.preventDefault();
      const { toggleTaskState, inverseTaskTheme } = this.props;
      inverseTaskTheme({ id });
      toggleTaskState({ id });
    };

    this.handleRemove = (id) => (event) => {
      event.preventDefault();
      const { removeTask } = this.props;
      removeTask({ id });
    };
  }

  render() {
    const { tasks, tasksUIState } = this.props;
    const applyTheme = (theme) => {
      const regularStyle = 'list-group-item d-flex';
      const themeClasses = {
        light: 'bg-light text-dark',
        dark: 'bg-dark text-light',
      };
      return cn(regularStyle, themeClasses[theme]);
    };

    if (tasks.length === 0) {
      return null;
    }

    return (
      <div className="mt-3">
        <ul className="list-group">
          {tasks.map(({ id, text, state }) => (
            <li className={applyTheme(tasksUIState[id].theme)} key={id}>
              <span className="mr-auto">
                <a href="#" data-test="task-toggle-state" onClick={this.handleToggleTaskState(id)}>
                  {state === 'active' ? text : <s>{text}</s>}
                </a>
              </span>
              <button type="button" className="close" onClick={this.handleRemove(id)}>
                <span>&times;</span>
              </button>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

export default connect(mapStateToProps, actionsList)(Tasks);
