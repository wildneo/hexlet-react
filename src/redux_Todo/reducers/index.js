import { combineReducers } from 'redux';

const text = (state = '', action) => {
  switch (action.type) {
    case 'TEXT_UPDATE':
      return action.payload.text;

    case 'TASK_ADD':
      return '';

    default:
      return state;
  }
};

const tasks = (state = [], action) => {
  switch (action.type) {
    case 'TASK_ADD': {
      const { task } = action.payload;
      return [task, ...state];
    }

    case 'TASK_REMOVE': {
      const { id } = action.payload;
      return state.filter((task) => task.id !== id);
    }

    case 'TASK_CLEAN':
      return [];

    default:
      return state;
  }
};

export default combineReducers({
  text,
  tasks,
});
