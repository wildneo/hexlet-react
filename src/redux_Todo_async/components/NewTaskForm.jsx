import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions';

const mapStateToProps = (state) => {
  const props = {
    text: state.text,
  };
  return props;
};

const actionCreators = {
  addTask: actions.addTask,
};

class NewTaskForm extends React.Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit({ text }) {
    const { addTask, reset } = this.props;
    addTask({ text });
    reset();
  }

  render() {
    const {
      handleSubmit, submitting, pristine,
    } = this.props;

    return (
      <form className="form-inline" onSubmit={handleSubmit(this.handleSubmit)}>
        <div className="form-group mx-3">
          <Field name="text" required component="input" type="text" />
        </div>
        <input type="submit" disabled={pristine || submitting} className="btn btn-primary btn-sm" value="Add" />
      </form>
    );
  }
}

const ConnectedNewTaskForm = connect(mapStateToProps, actionCreators)(NewTaskForm);
export default reduxForm({
  form: 'newTask',
})(ConnectedNewTaskForm);
