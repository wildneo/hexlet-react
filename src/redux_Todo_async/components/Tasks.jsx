import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions';

const mapStateToProps = (state) => {
  const { tasksFetchingState, tasks: { byId, allIds } } = state;
  const tasks = allIds.map((id) => byId[id]);
  return { tasks, tasksFetchingState };
};

const actionCreators = {
  removeTask: actions.removeTask,
};

class Tasks extends React.Component {
  constructor(props) {
    super(props);

    this.handleRemoveTask = (id) => () => {
      const { removeTask } = this.props;
      removeTask({ id });
    };
  }

  render() {
    const { tasks, tasksFetchingState } = this.props;

    if (tasksFetchingState === 'requested') {
      return (
        <div className="spinner-border m-3" role="status">
          <span className="sr-only">Loading...</span>
        </div>
      );
    }
    if (tasksFetchingState === 'failed') {
      return (
        <span>Please, reload page!</span>
      );
    }

    if (tasks.length === 0) {
      return null;
    }

    return (
      <div className="mt-3">
        <ul className="list-group">
          {tasks.map(({ id, text }) => (
            <li key={id} className="list-group-item d-flex">
              <span className="mr-auto">{text}</span>
              <button type="button" data-test="task-remove" className="close" onClick={this.handleRemoveTask(id)}>
                <span>&times;</span>
              </button>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

export default connect(mapStateToProps, actionCreators)(Tasks);
