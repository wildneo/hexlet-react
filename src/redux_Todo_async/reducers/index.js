import _ from 'lodash';
import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import { reducer as formReducer } from 'redux-form';
import * as actions from '../actions';

// BEGIN (write your solution here)
const taskRemovingState = handleActions({
  [actions.fetchTasksRequest]: () => 'requested',
  [actions.fetchTasksFailure]: () => 'failed',
  [actions.fetchTasksSuccess]: () => 'finished',
}, 'none');
// END

const tasksFetchingState = handleActions({
  [actions.fetchTasksRequest]: () => 'requested',
  [actions.fetchTasksFailure]: () => 'failed',
  [actions.fetchTasksSuccess]: () => 'finished',
}, 'none');

const tasks = handleActions({
  [actions.fetchTasksSuccess]: (state, { payload }) => ({
    byId: _.keyBy(payload.tasks, 'id'),
    allIds: payload.tasks.map((t) => t.id),
  }),

  [actions.addTaskSuccess]: (state, { payload: { task } }) => {
    const { byId, allIds } = state;

    return {
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },

  [actions.removeTaskSuccess]: (state, { payload: { id } }) => {
    const { byId, allIds } = state;

    return {
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
    };
  },
}, { byId: {}, allIds: [] });

export default combineReducers({
  taskRemovingState,
  tasksFetchingState,
  tasks,
  form: formReducer,
});
