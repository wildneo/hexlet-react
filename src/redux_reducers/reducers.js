import _ from 'lodash';
import { combineReducers } from 'redux';

const tasks = (state = {}, action) => {
  switch (action.type) {
    case 'TASK_ADD': {
      const { task } = action.payload;

      return { ...state, [task.id]: task };
    }

    case 'TASK_REMOVE': {
      const { id } = action.payload;

      return _.omit(state, id);
    }

    default:
      return state;
  }
};

const comments = (state = {}, action) => {
  switch (action.type) {
    case 'TASK_REMOVE': {
      const { id } = action.payload;

      return _.omitBy(state, (c) => c.taskId === id);
    }

    case 'TASK_COMMENT_ADD': {
      const { comment } = action.payload;

      return { ...state, [comment.id]: comment };
    }

    case 'TASK_COMMENT_REMOVE': {
      const { id } = action.payload;

      return _.omit(state, id);
    }

    default:
      return state;
  }
};

export default combineReducers({ tasks, comments });
