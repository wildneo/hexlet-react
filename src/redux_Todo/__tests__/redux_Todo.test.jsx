import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import reducers from '../reducers';
import NewTaskForm from '../components/NewTaskForm';
import Tasks from '../components/Tasks';

Enzyme.configure({ adapter: new Adapter() });

test('Store', () => {
  const store = createStore(reducers);

  const vdom = (
    <Provider store={store}>
      <div className="col-5">
        <NewTaskForm />
        <Tasks />
      </div>
    </Provider>
  );
  const wrapper = mount(vdom);
  expect(wrapper.render()).toMatchSnapshot();

  const newTaskInput = wrapper.find('input[type="text"]');
  const newTaskSubmit = wrapper.find('button[type="submit"]');

  newTaskInput.simulate('change', { target: { value: 'na-na' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'another task' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  const pressCloseTask1 = wrapper.find('.close').first();
  pressCloseTask1.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'na-na' } });
  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  const pressCloseTask2 = wrapper.find('.close').first();
  pressCloseTask2.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
