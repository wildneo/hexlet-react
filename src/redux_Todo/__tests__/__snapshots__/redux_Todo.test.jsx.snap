// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`Store 2`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value="na-na"
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`Store 3`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 4`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value="another task"
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 5`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          another task
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 6`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 7`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Store 8`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3"
  >
    <ul
      class="list-group"
    >
      <li
        class="list-group-item d-flex"
      >
        <span
          class="mr-auto"
        >
          na-na
        </span>
        <button
          class="close"
          type="button"
        >
          <span>
            ×
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
